### PR TITLE
Zoltan2: Fix rectilinear running non-rectilinear code for MJ (#7351)

### LINKED ISSUE #7351: zoltan2:  memory error when multijagged generates rectilinear partitions
## Bug Report
@trilinos/zoltan2 @kuberry 

### Description
When generating rectilinear partitions, multijagged appears to have a memory error writing to local_cut_line_weight_to_put_left.

https://github.com/trilinos/Trilinos/blob/0c5b9ccf004d4efee9475d5aaef9139f7be2044d/packages/zoltan2/src/algorithms/partition/Zoltan2_AlgMultiJagged.hpp#L4721

A check of the extent for one test case shows the view has length zero, but i == num_cuts-1 == 1.

I am wondering if the bug is here, as the comment and the if-test differ:
https://github.com/trilinos/Trilinos/blob/0c5b9ccf004d4efee9475d5aaef9139f7be2044d/packages/zoltan2/src/algorithms/partition/Zoltan2_AlgMultiJagged.hpp#L2636

It appears thread_cut_line_weight_to_put_left should always be allocated, not only when not generating rectilinear parts. I don't understand the code enough, though.  Do the arrays *_rectilinear_cut_weight get used only when generating rectilinear parts?  Or only when NOT generating rectilinear parts?

@MicheldeMessieres  and I will discuss.

### Steps to Reproduce

On cee-lan:  module load sierra-devel/gcc-8.3.0-openmpi-1.10.2

```
cmake \
\
-D Trilinos_ENABLE_EXPLICIT_INSTANTIATION:BOOL=ON \
-D Tpetra_INST_INT_LONG_LONG:BOOL=ON \
-D Tpetra_INST_INT_INT:BOOL=OFF \
-D Tpetra_INST_DOUBLE:BOOL=ON \
-D Tpetra_INST_FLOAT:BOOL=OFF \
-D Tpetra_INST_COMPLEX_FLOAT:BOOL=OFF \
-D Tpetra_INST_COMPLEX_DOUBLE:BOOL=OFF \
-D Tpetra_INST_INT_LONG:BOOL=OFF \
-D Tpetra_INST_INT_UNSIGNED:BOOL=OFF \
-D Tpetra_INST_SERIAL:BOOL=ON \
\
-D CMAKE_BUILD_TYPE:STRING=DEBUG  \
-D CMAKE_VERBOSE_MAKEFILE:BOOL=OFF \
-D CMAKE_C_FLAGS:STRING="-fopenmp -Wno-sign-compare" \
-D CMAKE_CXX_FLAGS:STRING="-fopenmp -Wno-sign-compare" \
\
-D TPL_ENABLE_MPI:BOOL=ON \
-D MPI_EXEC_MAX_NUMPROCS:STRING=11 \
\
-D TPL_ENABLE_Pthread:BOOL=OFF \
\
-D Trilinos_ENABLE_Kokkos:BOOL=ON \
-D Trilinos_ENABLE_KokkosCore:BOOL=ON \
-D Trilinos_ENABLE_KokkosAlgorithms:BOOL=ON \
-D Kokkos_ENABLE_Serial:BOOL=ON \
\
-D Trilinos_ENABLE_ALL_OPTIONAL_PACKAGES:BOOL=OFF \
-D Trilinos_ENABLE_TESTS:BOOL=OFF \
-D Trilinos_ENABLE_EXAMPLES:BOOL=OFF \
-D Trilinos_VERBOSE_CONFIGURE:BOOL=OFF \
\
-D Trilinos_ENABLE_Zoltan2:BOOL=ON \
-D Zoltan2_ENABLE_EXAMPLES:BOOL=OFF \
-D Zoltan2_ENABLE_TESTS:BOOL=ON \
\
-D Teuchos_ENABLE_STACKTRACE=OFF
```

In branch https://github.com/trilinos/Trilinos/tree/zoltan2_kuberry, 
the test zoltan2/test/driver/driver_inputs/kuberry will show an invalid write in valgrind:
valgrind Zoltan2_test_driver.exe kuberry.xml

### PR DESCRIPTION
**Pull request by MicheldeMessieres**
Resolves #7351 

@trilinos/zoltan2

## Motivation

Fixes incorrect code in MJ which was not supposed to run for rectilinear.
Adds a test to cover the situation.

## Testing

Adds a new kuberry test to cover this situation.

### CODE REVIEW

File: packages/zoltan2/test/driver/driverinputs/kuberry/kuberry [hunk 1]
```diff
@@ -0,0 +1,21 @@
+# Test problem that exposes a memory leak in MJ when num_global_parts==1
```
Comment (kddevin) [inline packages/zoltan2/test/driver/driverinputs/kuberry/kuberry:1]: This file isn't necessary, is it?

File: packages/zoltan2/test/driver/driverinputs/kuberry/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,21 @@
+# Test problem that exposes a memory leak in MJ when num_global_parts==1
```
Comment (kddevin) [inline packages/zoltan2/test/driver/driverinputs/kuberry/CMakeLists.txt:1]: Correct comment:
"Test problem for #7351 that exposes an out-of-bounds array access in MJ when rectilinear = 1"

File: packages/zoltan2/test/driver/driverinputs/kuberry/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+TRIBITS_ADD_TEST(
+    test_driver
+    NAME kuberry
+    DIRECTORY ../..
+    NUM_MPI_PROCS 4
```
Comment (kddevin) [inline packages/zoltan2/test/driver/driverinputs/kuberry/CMakeLists.txt:6]: No need to specify num_mpi_procs; OK to run in serial, too.

File: packages/zoltan2/test/driver/driverinputs/kuberry/CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+    "kuberry.xml"
+    PASS_REGULAR_EXPRESSION "PASS"
+    FAIL_REGULAR_EXPRESSION "FAIL"
+)
+
+TRIBITS_COPY_FILES_TO_BINARY_DIR(copy_kuberry
+    SOURCE_FILES
+        kuberry.xml
+        kuberry.graph
+        kuberry.coords
```
Comment (kddevin) [inline packages/zoltan2/test/driver/driverinputs/kuberry/CMakeLists.txt:18]: need kuberry.assign here as well

File: packages/zoltan2/test/driver/driverinputs/kuberry/kuberry (hunk 1 above)
Comment (MicheldeMessieres) [reply packages/zoltan2/test/driver/driverinputs/kuberry/kuberry]: Removed

File: packages/zoltan2/test/driver/driverinputs/kuberry/CMakeLists.txt (hunk 2 above)
Comment (MicheldeMessieres) [reply packages/zoltan2/test/driver/driverinputs/kuberry/CMakeLists.txt]: Fixed

File: packages/zoltan2/test/driver/driverinputs/kuberry/CMakeLists.txt (hunk 3 above)
Comment (MicheldeMessieres) [reply packages/zoltan2/test/driver/driverinputs/kuberry/CMakeLists.txt]: Removed

File: packages/zoltan2/test/driver/driverinputs/kuberry/CMakeLists.txt (hunk 4 above)
Comment (MicheldeMessieres) [reply packages/zoltan2/test/driver/driverinputs/kuberry/CMakeLists.txt:18]: Added

**Pull request merged**